### DISCUSSION TITLE
Why are the libraries always re-build?

### DISCUSSION THREAD

**Discussion by rempas** [vlang/v#13680] [General]
I'm trying to built the [calculator](https://github.com/vlang/ui/blob/master/examples/calculator.v) example from the [ui](https://github.com/vlang/ui) library and for a production build, it takes about 19 seconds on my machine. This is 80% slower than [ved](https://github.com/vlang/ved) and the source code of the example is 16 times less so I'm pretty sure that V always builds "ui" too.

But why is this happening? It doesn't make much sense as the library will never change from me. How can I have the library to not automatically be build every time?

**Comment by changrui** [comment]
`use-cache`?

**Reply by rempas** [reply]
I'm trying to compile using: `v test.v -use-cache -prod` and the result is the same. Unless I didn't understand what you are talking about.

**Reply by changrui** [reply]
oh, no, I'm sorry. The option is `-usecache`. It has comsupt half time when I compile v self between `v -usecache -prod -o v cmd/v` and `v -prod -o v cmd/v`, so I guess v do not `re-compile` builtins.

**Reply by changrui** [reply]
It wil take much more time when the first use of `-usecache` option.

**Reply by rempas** [reply]
Thanks! Unfortunately, I got an error when trying to use this option so I cannot tell you how effective it is for my case.

I will create a new issue.

Also, using "-use-cache" doesn't seem to generate a compiler error regardless if the option is not accepted. Have you crossed in any discussion or issue talking about that and why that is? I'm thinking of making an issue for that too.

**Reply by spytheman** [reply]
All V options should go *before* the source file/folder:
`v -usecache -prod test.v`

**Reply by spytheman** [reply]
Try for example: `time v -usecache -prod test.v` 2 times - the first time, it will be slower (because it will compile all the needed modules separately), but the second time, it should be much faster (as long as you do not change the source code for the used modules).

**Reply by spytheman** [reply]
You can also do `v wipe-cache` to remove the accumulated cached objects, if you want to repeat your experiments.

**Reply by spytheman** [reply]
NB: `-usecache` is not yet the default, because some modules are not compatible yet with `-usecache` (mainly the ones that depend on generics like `datatype` or `math` for example).

**Reply by rempas** [reply]
@spytheman Do you know if `-usecache` creates the library/executable for the library in a global scope? If I create another project, will I have to re-create the cache for that library the first time that I will try to compile?

**Reply by spytheman** [reply]
The cache for the modules is global, so it will work for all programs that use the same module (like `builtin` for example is used by every V program).

The cache keys however are dependent on the used CLI options, i.e. 
`v -usecache -prod file.v`,
... will use a different cache entry for each module, than:
`v -usecache file.v`
... and a different for:
`v -usecache -cflags -O2 file.v`

**Reply by rempas** [reply]
That's awesome! From what you said: 

> NB: `-usecache` is not yet the default, because some modules are not compatible yet with `-usecache` (mainly the ones that depend on generics like `datatype` or `math` for example).

I suppose that you know that this will become the default in the future when it becomes stable?

**Reply by spytheman** [reply]
Yes, afaik it will be on by default, when it improves to work for all V code, including code that uses generics.

**Reply by rempas** [reply]
That's awesome! Thanks for your time!
